install-filecoin

```diff
@@ -2,12 +2,137 @@
+    # function's caller provide the release name.
+    __release_name="${__repo_name}-$(uname)-${__release_type}"
+    __release_sha1=$(git rev-parse HEAD)
+    __release_tag="${__release_sha1:0:16}"
+    __release_tag_url="https://api.github.com/repos/filecoin-project/${__repo_name}/releases/tags/${__release_tag}"
+
+    (>&2 echo "acquiring release @ ${__release_tag}")
+
+    __release_response=$(curl \
+        --retry 3 \
+        --location $__release_tag_url)
+
+    __release_url=$(echo $__release_response | jq -r ".assets[] | select(.name | contains(\"${__release_name}\")) | .url")
+
+    if [[ -z "$__release_url" ]]; then
+        (>&2 echo "failed to download release (tag URL: ${__release_tag_url}, response: ${__release_response})")
+        return 1
+    fi
+
+    __tar_path="/tmp/${__release_name}_$(basename ${__release_url}).tar.gz"
+
+    __asset_url=$(curl \
+        --head \
+        --retry 3 \
+        --header "Accept:application/octet-stream" \
+        --location \
+        --output /dev/null \
+        -w %{url_effective} \
+        "$__release_url")
+
+    curl --retry 3 --output "${__tar_path}" "$__asset_url"
+    if [[ $? -ne "0" ]]; then
+        (>&2 echo "failed to download release asset (tag URL: ${__release_tag_url}, asset URL: ${__asset_url})")
+        return 1
+    fi
+
+    eval $__resultvar="'$__tar_path'"
+}
+
+build_from_source() {
+    __library_name=$1
+    __rust_sources_path=$2
+    __release_type=$3
+    __repo_sha1=$(git rev-parse HEAD)
+    __repo_sha1_truncated="${__repo_sha1:0:16}"
+
+    (>&2 echo "building from source @ ${__repo_sha1_truncated}")
+
+    if ! [ -x "$(command -v cargo)" ]; then
+        (>&2 echo 'Error: cargo is not installed.')
+        (>&2 echo 'Install Rust toolchain to resolve this problem.')
+        exit 1
+    fi
+
+    if ! [ -x "$(command -v rustup)" ]; then
+        (>&2 echo 'Error: rustup is not installed.')
+        (>&2 echo 'Install Rust toolchain installer to resolve this problem.')
+        exit 1
+    fi
+
+    pushd $__rust_sources_path
+
+    cargo --version
+
+    target_feature=""
+    for x in ${req_cpu_features_optimized[@]}; do
+        target_feature="+${x},${target_feature}"
+    done
+
+    if [ "${__release_type}" = "optimized" ]; then
+        RUSTFLAGS="-C target-feature=${target_feature}" ./scripts/build-release.sh $__library_name $(cat rust-toolchain)
+    else
+        ./scripts/build-release.sh $__library_name $(cat rust-toolchain)
+    fi
+
+    popd
+}
+
+get_release_type() {
+    searched="cat /proc/cpuinfo | grep flags"
+    features=""
+
+    # determine where to look for CPU features
+    #
+    if [[ ! -f "/proc/cpuinfo" ]]; then
+        (>&2 echo "No /proc/cpuinfo file; falling back to Darwin feature detection")
+        searched="sysctl -a | grep machdep.cpu | tr '[:upper:]' '[:lower:]' | grep features"
+        features=$(eval "$searched")
+    else
+        features=$(cat /proc/cpuinfo | grep flags)
+    fi
```

Comment:
Your code works, but I would write it as

```suggestion
    searched=""

    # determine where to look for CPU features
    #
    if [[ -f "/proc/cpuinfo" ]]; then
        searched="cat /proc/cpuinfo | grep flags"
    else
        (>&2 echo "No /proc/cpuinfo file; falling back to Darwin feature detection")
        searched="sysctl -a | grep machdep.cpu | tr '[:upper:]' '[:lower:]' | grep 
    fi
    features=$(eval "${searched}")
```

Reply:
@vmx 

> Your code works, but I would write it as

good idea :)